### DISCUSSION TITLE
GotoState

### DISCUSSION THREAD

**Discussion by winksaville** [AletheiaWareLLC/Wink#22] [Ideas]
I was looking at [GotoState](https://github.com/AletheiaWareLLC/Wink/blob/164c7b681cd9b36edde05416c099ab67f7a2b69f/src/machine.cpp#L100) AFAICT it only handles an HSM consisting of a parent and a child. I think eventually you'll need to handle more "complex" hierarchies, like:
```
Simple tree:
       parent
       /    \
      /      \
    leaf1   leaf2

Bigger tree:
         parent
         /    \
        /      \
      leaf1    child1
               /   \
              /     \
            leaf2  leaf3

Forest:
         parent1            leaf1            parent2                 parent3
         /    \                              /    \                     |
        /      \                            /      \                    |
      leaf2    child2                    child3    child4             leaf3
               /   \                      /          \
              /     \                    /            \
           leaf4   leaf5              leaf6         leaf7 
```

To do that you'll probably need to do more work in "handleMessage" and less in "GotoState". In my implementation of "transitioning" and "entry/exit" is done in [`dispatch_idx`](https://github.com/winksaville/proc-macro-hsm1/blob/7e4d143cf0ff1d9dfcdea06ffa4f4bd441474ecf/hsm0-with-executor/src/lib.rs#L341) which is equivalent to your "GotoState" and "handleMessage" routines. Although I don't have a "GotoState" having one works just as well, but to handle more complex hierarchies I think it would probably just save the "target_state" into the state object and then "handleMessage" would do the entry/exit code. Note; if you'd like to go over `dispatch_idx` or anything else in my Rust HSM it we can do a "video" call.

The other thing I did in my Rust HSM is I only allow transitioning between leafs. I made this decision for two reasons:

 1) I contend that if you allow a "child" or "parent" to be the current state than it must also process all messages that are explicitly handled by it's children. This means there "must" be code duplication.

 2) I couldn't decide what the "right" solution was for some transitions. For instance, in "Simple Tree" a transition from leaf1 to leaf2. You definitely exit leaf1 and enter leaf2 but do you exit and enter parent? I've argued with my self both ways. Also, what about a transition from leaf2 to parent? You definitely exit leaf2 but do you enter parent? Again it's debatable.

Therefore, for me, reason 1 is enough to sway me to decide to only allow transitions between leafs. Looking forward to hearing your opinion.

**Comment by stuartmscott** [comment]
I've created https://github.com/AletheiaWareLLC/Wink/pull/24 to codify the example tree shapes you gave.

GotoState will execute the current state's `OnExit` action, and the next state's `OnEntry` action, but then it will return and continue execution of the action that triggered the state change, which could be an issue if the next statement isn't a `return`.

```
GotoState("foo");
GotoState("bar");
return;
```

The current state's `OnExit` will be executed followed by `foo.OnEntry`, then `foo.OnExit`, then `bar.OnEntry`. Ideally, calling `GotoState` would immediately stop executing the handler, transition to the next state, and await the next message.

> only allow transitioning between leafs

I haven't added this restriction, no sure it is necessary.

> must also process all messages that are explicitly handled by it's children

I haven't added this restriction either, though I do recognize there is value in being able to quantify the set of messages a SM will handle and define the interface/protocol it implements. Currently these SMs will error if a message is received that it can't handle, it doesn't have a generic handler for, and its hierarchy can't handle. This error will get propagated to the spawner.

> code duplication

If the root state handles all messages that the lower levels can, the implementation will likely be a `safe default` and share little-to-no code with the `actual operation` in the leaf.

> 2.

Interesting debate. One the one hand these states are nodes in a tree, on the other they are nested boxes. Either exit one state and immediately enter another, or traverse up the tree exiting all states along the way to the common ancestor then down the tree entering all states along the way. At the moment this framework does the former, but I'm leaning towards the latter.

**Comment by stuartmscott** [comment]
Created https://github.com/AletheiaWareLLC/Wink/issues/45 to track how transitions should traverse the hierarchy.
